http: add CONNECT-UDP support

jeongseokson opened the pull request:
Commit Message:
This commit adds CONNECT-UDP (RFC 9298) support. UdpConnPool is added to create a UDP socket for a new CONNECT-UDP request, and UDPUpstream is added to maintain the socket and other relevant data associated with UDP upstreams. 

We added an integration test for the terminating CONNECT-UDP proxy, but not the forwarding proxy in this commit. We are going to add test cases to cover the forwarding proxy scenario in a subsequent commit. 

Additional Description:
Risk Level: Medium, the feature can only be enabled by the new configuration added in this commit.
Testing: Integration test
Runtime guard: envoy.reloadable_features.enable_connect_udp_support
Release Notes:  added support for CONNECT-UDP (RFC 9298). Can be disabled by setting runtime feature ``envoy.reloadable_features.enable_connect_udp_support`` to false.

jeongseokson commented:
/assign @DavidSchinazi

David will do a first pass review. Thank you!

jeongseokson commented:
What's the right way to add api/envoy/extensions/upstreams/http/udp/v3/udp_connection_pool.proto? The check_and_fix_proto_format CI complains about it.

silverstar194 commented on source/common/http/conn_manager_impl.cc:
```diff
@@ -1167,6 +1167,14 @@ void ConnectionManagerImpl::ActiveStream::decodeHeaders(RequestHeaderMapSharedPt
     return;
   }
 
+  // Rewrites the host of CONNECT-UDP requests.
```
Do these need to be guarded with a flag?

jeongseokson replied:
That would probably a safer choice. I would like to know other reviewers' thoughts on this as well. Thanks for your comment, Sean.

jeongseokson replied:
Done.

tyxia commented:
> What's the right way to add api/envoy/extensions/upstreams/http/udp/v3/udp_connection_pool.proto? The check_and_fix_proto_format CI complains about it.

Drive-by comment.... The proto BUILD file is generated by tool. Once you have the correct proto change (e.g., import, namespace, etc), you can run `fix_proto_format`. I used `sudo ./ci/run_envoy_docker.sh '\''./ci/do_ci.sh fix_proto_format` locally.

jeongseokson commented:
> > What's the right way to add api/envoy/extensions/upstreams/http/udp/v3/udp_connection_pool.proto? The check_and_fix_proto_format CI complains about it.
> 
> Drive-by comment.... The proto BUILD file is generated by tool. Once you have the correct proto change (e.g., import, namespace, etc), you can run `fix_proto_format`. I used `sudo ./ci/run_envoy_docker.sh '\''./ci/do_ci.sh fix_proto_format` locally.

I tried it and it deleted both the proto and the BUILD file. Tried it with the proto file alone then it complained about missing the BUILD file. I guess I should have written the proto in a different way? I don't see any Envoy documentation about this.

DavidSchinazi commented on source/common/http/header_utility.cc:
```diff
@@ -20,6 +20,15 @@
 
+namespace {
+
+const Regex::CompiledGoogleReMatcher& connectUdpPathMatcher() {
```
A regex seems pretty heavyweight for something this simple. I'd assume we'd get noticeably better performance by checking whether the path is prefixed by `/.well-known/masque/udp/`, removing the prefix and then splitting on `/`.

That said, this could be a premature optimization. Is there precedent for regex vs split in the codebase currently?

jeongseokson replied:
Regex is used in Envoy mainly for rewriting headers based on the user configuration. I agree that in this case, regex could be an overkill. I think this could make more sense in a later stage if we want to support more custom URI Templates. I will remove regex as you suggested.

DavidSchinazi commented on source/common/http/header_utility.cc:
```diff
@@ -243,13 +252,28 @@ bool HeaderUtility::isConnect(const RequestHeaderMap& headers) {
              Http::Code::OK;
 }
 
+bool HeaderUtility::rewriteAuthorityForConnectUdp(RequestHeaderMap& headers) {
```
This function has two issues:
1. it's missing the required percent-decoding
2. it's not adding the `[` and `]` required when using IPv6 addresses in URLs (see definition of [`authority`](https://www.rfc-editor.org/rfc/rfc3986#section-3.2))

Both of these would cause IPv6 literals to fail, are those covered in tests?

jeongseokson replied:
Thank you for catching them. I will send an additional commit soon. One problem of the existing test cases is that it doesn't really make use of the parsed host and port. Envoy always just connects to a configured test cluster. I reality, a Dynamic Forward Proxy Filter & Cluster should be configured for this. I will think about some test options and cover it in the test cases.

jeongseokson replied:
I fixed the bug and added relevant unit test and integration test cases. Done.

DavidSchinazi commented on source/common/quic/envoy_quic_client_session.h:
```diff
@@ -65,10 +65,11 @@ class EnvoyQuicClientSession : public QuicFilterManagerConnectionImpl,
-    // TODO(https://github.com/envoyproxy/envoy/issues/23564): Http3 Datagram support should be
-    // turned on by returning quic::HttpDatagramSupport::kRfc once the CONNECT-UDP support work is
-    // completed.
+#ifdef ENVOY_ENABLE_HTTP_DATAGRAMS
```
Should this be a runtime guard instead of compile time? Ideally the presence of CONNECT-UDP in the config should enable this. Same comment for all other uses of this macro in this PR

jeongseokson replied:
Added the runtime guard.

DavidSchinazi replied:
What's the purpose of removing this from Envoy Mobile at compile time? They're tight on binary size but this is something that will be needed there eventually and a macro adds complexity

DavidSchinazi commented on source/extensions/upstreams/http/generic/config.cc:
```diff
@@ -2,26 +2,39 @@
+    auto udp_conn_pool =
+        std::make_unique<Upstreams::Http::Udp::UdpConnPool>(thread_local_cluster, ctx);
+    return (udp_conn_pool->valid() ? std::move(udp_conn_pool) : nullptr);
+  }
```
Add `return nullptr` after the `switch` to avoid returning uninitialized memory if `upstream_protocol` is invalid

DavidSchinazi commented on source/extensions/upstreams/http/udp/upstream_request.cc:
```diff
@@ -0,0 +1,141 @@
+      Event::FileReadyType::Read);
+}
+
+void UdpUpstream::encodeData(Buffer::Instance& data, bool /*end_stream*/) {
```
What do `data` and `end_stream` represent here? I was assuming that `data` was a chunk of the capsule stream and `end_stream` means the capsule stream is over, but if that's the case we should only call `capsule_parser_.ErrorIfThereIsRemainingBufferedData()` once `end_stream` is set

DavidSchinazi commented on source/extensions/upstreams/http/udp/upstream_request.cc:
```diff
@@ -0,0 +1,141 @@
+      Envoy::Http::createHeaderMap<Envoy::Http::ResponseHeaderMapImpl>(
+          {{Envoy::Http::Headers::get().Status, "200"},
+           {Envoy::Http::Headers::get().CapsuleProtocol, "?1"}})};
+  upstream_to_downstream_->decodeHeaders(std::move(headers), false);
```
nit: `/*variable_name=*/false`

DavidSchinazi replied:
Doesn't appear done?

DavidSchinazi commented on source/extensions/upstreams/http/udp/upstream_request.cc:
```diff
@@ -0,0 +1,141 @@
+  uint32_t packets_dropped = 0;
+  const Api::IoErrorPtr result = Network::Utility::readPacketsFromSocket(
+      socket_->ioHandle(), *socket_->connectionInfoProvider().localAddress(), *this,
+      dispatcher_.timeSource(), /*prefer_gro=*/false, packets_dropped);
```
Why is `prefer_gro` false?

DavidSchinazi commented on source/extensions/upstreams/http/udp/upstream_request.cc:
```diff
@@ -0,0 +1,141 @@
+  }
+}
+
+void UdpUpstream::processPacket(Network::Address::InstanceConstSharedPtr /*local_address*/,
```
To confirm: we're ignoring `local_address` and `peer_address` here because we're using a bound and connected socket?

DavidSchinazi commented on source/extensions/upstreams/http/udp/upstream_request.cc:
```diff
@@ -0,0 +1,141 @@
+bool UdpUpstream::OnCapsule(const quiche::Capsule& capsule) {
+  quiche::CapsuleType capsule_type = capsule.capsule_type();
+  if (capsule_type != quiche::CapsuleType::DATAGRAM) {
+    // Silently drops Datagram Capsules with an unknown type.
```
nit: these aren't datagram capsules
```suggestion
    // Silently drops capsules with an unknown type.
```

vparla replied:
Can we get datagram capsules support?  this is critical to the implementation for Zero Trust Access use cases.

DavidSchinazi commented on source/extensions/upstreams/http/udp/upstream_request.cc:
```diff
@@ -0,0 +1,141 @@
+  Buffer::InstancePtr buffer = std::make_unique<Buffer::OwnedImpl>();
+  buffer->add(connect_udp_datagram_payload->GetUdpProxyingPayload());
+  bytes_meter_->addWireBytesSent(buffer->length());
+  Api::IoCallUint64Result rc = Network::Utility::writeToSocket(
```
Are we ignoring socket write errors? If so, can we add a TODO to react to some errors? For example if we get "no route to host" we could close the stream. Similarly, I notice that this file tracks the number of dropped datagrams, we could increase that number when we get socket write errors

jeongseokson replied:
Good point. I added a TODO. The existing CONNECT implementation doesn't keep track of errors so perhaps I could enhance the CONNECT implementation together.

DavidSchinazi replied:
SGTM

DavidSchinazi commented on source/extensions/upstreams/http/udp/upstream_request.h:
```diff
@@ -0,0 +1,105 @@
+
+  Network::SocketPtr createSocket(const Upstream::HostConstSharedPtr& host) {
+    return std::make_unique<Network::SocketImpl>(Network::Socket::Type::Datagram, host->address(),
+                                                 nullptr, Network::SocketCreationOptions{});
```
nit: `/*variable_name=*/nullptr`

DavidSchinazi commented on test/integration/udp_tunneling_integration_test.cc:
```diff
@@ -0,0 +1,243 @@
+  std::string received_capsule_fragment =
+      absl::HexStringToBytes("00"               // DATAGRAM capsule type
+                             "08"               // capsule length
+                             "a1a2a3a4a5a6a7a8" // HTTP Datagram payload
```
Why doesn't this have the `00` byte of the contextID=0?

DavidSchinazi reviewed (commented):
Thanks, this is overall in great shape!

jeongseokson commented on envoy/router/router.h:
```diff
@@ -1515,14 +1515,20 @@ using GenericConnPoolPtr = std::unique_ptr<GenericConnPool>;
   virtual GenericConnPoolPtr
-  createGenericConnPool(Upstream::ThreadLocalCluster& thread_local_cluster, bool is_connect,
+  createGenericConnPool(Upstream::ThreadLocalCluster& thread_local_cluster,
+                        GenericConnPoolFactory::UpstreamProtocol is_connect,
```
RIght, done.

jeongseokson reviewed (commented):
Thanks a lot for reviewing the PR, David. I will send another commit to complete the fix.

jeongseokson reviewed (commented):
Resolved the rest of the issues such as IPv6 parsing. Can you please review my new changes, David? Thank you.

DavidSchinazi commented on source/common/http/header_utility.cc:
```diff
@@ -243,13 +261,39 @@ bool HeaderUtility::isConnect(const RequestHeaderMap& headers) {
+
+  // Utility::PercentEncoding::decode never returns an empty string if the input argument is not
+  // empty.
+  std::string target_host = Utility::PercentEncoding::decode(path_split[4]);
```
Per the spec: `Note that IPv6 scoped addressing zone identifiers are not supported.`

We should return false if `target_host` contains a `%` after percent decoding

DavidSchinazi commented on source/extensions/upstreams/http/udp/upstream_request.cc:
```diff
@@ -0,0 +1,141 @@
+}
+
+Envoy::Http::Status UdpUpstream::encodeHeaders(const Envoy::Http::RequestHeaderMap&,
+                                               bool /*end_stream*/) {
```
Per the spec `The lifetime of the socket is tied to the request stream` so if we get a FIN on the stream then we should tear everything down. It would be silly for the client to FIN the stream right after the headers but if it happens we should tear everything down and not bother opening the socket at all. Maybe reply with status code 400?

jeongseokson replied:
That makes sense. The end_stream value should be passed to the decodeHeaders for cleaning up if that happens. I modified the code accordingly.

DavidSchinazi commented on source/extensions/upstreams/http/udp/upstream_request.cc:
```diff
@@ -0,0 +1,147 @@
+  return Envoy::Http::okStatus();
+}
+
+void UdpUpstream::resetStream() { upstream_to_downstream_ = nullptr; }
```
Should this close the socket or is that handled by the destructor?

DavidSchinazi commented on source/extensions/upstreams/http/udp/upstream_request.cc:
```diff
@@ -0,0 +1,147 @@
+  Buffer::InstancePtr capsule_data = std::make_unique<Buffer::OwnedImpl>();
+  capsule_data->add(serialized_capsule.AsStringView());
+  bytes_meter_->addWireBytesReceived(capsule_data->length());
+  upstream_to_downstream_->decodeData(*capsule_data, false);
```
nit: `/*variable_name=*/false`

DavidSchinazi commented on source/extensions/upstreams/http/udp/upstream_request.h:
```diff
@@ -0,0 +1,108 @@
+namespace Http {
+namespace Udp {
+
+class UdpConnPool : public Router::GenericConnPool {
```
nit: class comment, here and below

jeongseokson replied:
Done.

DavidSchinazi commented on test/common/http/header_utility_test.cc:
```diff
@@ -1148,6 +1148,15 @@ TEST(HeaderIsValidTest, IsConnect) {
+TEST(HeaderIsValidTest, IsConnectUdp) {
+  EXPECT_TRUE(
+      HeaderUtility::isConnectUdp(Http::TestRequestHeaderMapImpl{{"upgrade", "connect-udp"}}));
+  // Extended CONNECT requests should be normalied to HTTP/1.1.
```
```suggestion
  // Extended CONNECT requests should be normalized to HTTP/1.1.
```

jeongseokson replied:
Done.

DavidSchinazi commented on test/common/http/header_utility_test.cc:
```diff
@@ -1160,6 +1169,35 @@ TEST(HeaderIsValidTest, IsConnectResponse) {
+  EXPECT_TRUE(HeaderUtility::rewriteAuthorityForConnectUdp(connect_udp_request));
+  EXPECT_EQ(connect_udp_request.getHostValue(), "foo.lyft.com:80");
+  TestRequestHeaderMapImpl connect_udp_request_ipv6{
+      {":path", "/.well-known/masque/udp/2001:0db8:85a3::8a2e:0370:7334/80/"},
```
Per the spec `if "target_host" contains an IPv6 literal, the colons (":") MUST be percent-encoded. For example, if the target host is "2001:db8::42", it will be encoded in the URI as "2001%3Adb8%3A%3A42".`

jeongseokson replied:
Done.

DavidSchinazi commented on source/common/http/header_utility.cc:
```diff
@@ -243,13 +261,39 @@ bool HeaderUtility::isConnect(const RequestHeaderMap& headers) {
+    return false;
+  }
+
+  // Utility::PercentEncoding::decode never returns an empty string if the input argument is not
```
Per the spec `if "target_host" contains an IPv6 literal, the colons (":") MUST be percent-encoded. For example, if the target host is "2001:db8::42", it will be encoded in the URI as "2001%3Adb8%3A%3A42".`

We should return false if `path_split[4]` contains a `:`

DavidSchinazi commented on test/integration/udp_tunneling_integration_test.cc:
```diff
@@ -0,0 +1,261 @@
+    const std::string sent_capsule_fragment =
+        absl::HexStringToBytes("00"               // DATAGRAM capsule type
+                               "08"               // capsule length
+                               "00a1a2a3a4a5a6a7" // UDP Proxying HTTP Datagram payload
```
```suggestion
                               "00"             // context ID
                               "a1a2a3a4a5a6a7" // UDP Proxying Payload
```

Same below

DavidSchinazi reviewed (commented):
Overall this looks great! Please send to wider review after addressing these last comments

jeongseokson reviewed (commented):
Done. Thanks!

jeongseokson commented:
/assign @RyanTheOptimist 
/assign @alyssawilk 

Hi Ryan and Alyssa. @DavidSchinazi has finished the first pass so can you review this PR? Also, I have a couple of questions.
- What is the right way to add an API proto and a BUILD file? I added `api/envoy/extensions/upstreams/http/udp/v3/udp_connection_pool.proto` and `api/envoy/extensions/upstreams/http/udp/v3/BUILD` manually but the CI complains about it.
-  I think I should update relevant documentation as well. Is updating https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/http/upgrades would be enough? I will appreciate it if you can point me to other documents I have to update as well if any.

Thank you!

KBaichoo commented:
Can you take a look  @RyanTheOptimist ? Thanks

RyanTheOptimist commented:
> Can you take a look @RyanTheOptimist ? Thanks

Oh shoot. Sorry! I dropped this. Will take a look.

RyanTheOptimist commented:
Looks like format needs to be fixed:
```
./test/integration/udp_tunneling_integration_test.cc:237:      absl::HexStringToBytes("01"             // DATAGRAM Capsule Yype
                                                                                                                           ^
./test/integration/udp_tunneling_integration_test.cc:247:      absl::HexStringToBytes("00"             // DATAGRAM Capsule Yype
                                                                                                                           ^
Checked 4609 file(s) and 125517 comment(s), found 2 error(s).
```

And also the proto check is failing:
https://dev.azure.com/cncf/envoy/_build/results?buildId=139704&view=logs&jobId=c95d62a5-6bb9-58f0-2a02-7575fd482a15&j=c95d62a5-6bb9-58f0-2a02-7575fd482a15&t=76c9b3e6-0d40-57e9-06f3-2219e9228a0a

jeongseokson commented:
Yype! Fixed the typo. Do you know what is the right way to add that proto without encountering that check error, Ryan? I asked two questions including this one in the previous comment https://github.com/envoyproxy/envoy/pull/27714#issuecomment-1582777188 I will appreciate it if you can shed light on it.

RyanTheOptimist commented on source/common/router/router.cc:
```diff
@@ -769,18 +769,25 @@ Filter::createConnPool(Upstream::ThreadLocalCluster& thread_local_cluster) {
+    if (method == Http::Headers::get().MethodValues.Connect) {
+      upstream_protocol = UpstreamProtocol::TCP;
+    }
+    if (Runtime::runtimeFeatureEnabled("envoy.reloadable_features.enable_connect_udp_support") &&
```
Can / should this `if` and the next be `else if`? I *think* that the three branches are mutually exclusive, but I'm not sure.

jeongseokson replied:
The existing if statements were quite complicated to me as well so I refactored them in the new commit. Thanks for pointing this out.

RyanTheOptimist reviewed (commented):
Looks pretty good. A few cosmetic comments. I'm deferring to David's detailed review of the UDP connect details, but seems fine from my perspective :)

RyanTheOptimist commented:
> Yype! Fixed the typo. Do you know what is the right way to add that proto without encountering that check error, Ryan? I asked two questions including this one in the previous comment [#27714 (comment)](https://github.com/envoyproxy/envoy/pull/27714#issuecomment-1582777188) I will appreciate it if you can shed light on it.

I think you need to update api/BUILD. Check out https://github.com/envoyproxy/envoy/pull/27256/files as an example

That being said, the API shepherd should know for sure. @lizan, it looks like you're the shepherd for this PR. Can you do the API review?

jeongseokson commented on source/common/http/header_utility.cc:
```diff
@@ -243,13 +243,58 @@ bool HeaderUtility::isConnect(const RequestHeaderMap& headers) {
+  std::vector<absl::string_view> path_split = absl::StrSplit(path, '/');
+  if (path_split.size() != 7 || !path_split[0].empty() || path_split[1] != ".well-known" ||
+      path_split[2] != "masque" || path_split[3] != "udp" || path_split[4].empty() ||
+      path_split[5].empty() || !path_split[6].empty()) {
```
Right! I agree it will be more readable. Done.

jeongseokson commented on source/common/quic/envoy_quic_client_session.h:
```diff
@@ -65,10 +65,15 @@ class EnvoyQuicClientSession : public QuicFilterManagerConnectionImpl,
+#ifdef ENVOY_ENABLE_HTTP_DATAGRAMS
+    if (Runtime::runtimeFeatureEnabled("envoy.reloadable_features.enable_connect_udp_support")) {
+      return quic::HttpDatagramSupport::kRfc;
+    } else {
```
Good catch! Done.

jeongseokson reviewed (commented):
Thanks for the comments, Ryan. Let me know if you have any further comments.

jeongseokson commented:
> I think you need to update api/BUILD. Check out https://github.com/envoyproxy/envoy/pull/27256/files as an example
> 
> That being said, the API shepherd should know for sure. @lizan, it looks like you're the shepherd for this PR. Can you do the API review?

Thank you for the information, Ryan. Let me know if you have any comments, @lizan.

RyanTheOptimist commented:
Looks like a real build error:
```

# Execution platform: //third_party/rbe_configs/config:platform
external/envoy/source/common/router/router.cc:777:43: error: reference to type 'const Envoy::Http::RequestHeaderMap' could not bind to an lvalue of type 'Http::RequestHeaderMap *'
        Http::HeaderUtility::isConnectUdp(downstream_headers_)) {
                                          ^~~~~~~~~~~~~~~~~~~
external/envoy/source/common/http/header_utility.h:177:52: note: passing argument to parameter 'headers' here
  static bool isConnectUdp(const RequestHeaderMap& headers);
                                                   ^
1 error generated.
```

RyanTheOptimist reviewed (commented):
LGTM, modulo fixing CI failures

jeongseokson commented:
Ah, my bad. Just fixed it. Thanks for the review, Ryan. I will resolve CI errors if there's any.

RyanTheOptimist commented:
Looks like format is failing.
```
CodeChecker ERROR [yamllint] changelogs/current.yaml (key-duplicates 217:3): duplication of key "change" in mapping
```
I think this is missing the `-area` clause.

And looks like the docs need to be tweaked too:

```
Did you forget to add 'envoy.upstreams.http.udp' to extensions_build_config.bzl, extensions_metadata.yaml, contrib_build_config.bzl, or contrib/extensions_metadata.yaml?
```

RyanTheOptimist commented:
```
//test/integration:udp_tunneling_integration_test                       TIMEOUT in 6 out of 16 in 915.0s
```

jeongseokson commented:
@RyanTheOptimist This is quite tricky to debug. When I ran the test locally, the timeouts did not happen. I found that it happens with a certain chance only when:

* The downstream protocol is HTTP/3
* "-c opt" compilation flag is set when testing.

I have been trying to triage the cause but it's been quite difficult since I have to pass -c opt option to reproduce the timeout, so the QUICHE does not print any DLOGs.

Do you have any advice or tips to debug this type of issue, Ryan? I think it will be helpful if I can locally modify QUICHE library for Envoy to print out some debugging logs.

lizan reviewed (commented):
/lgtm api

jeongseokson commented:
Thank you, @lizan.

@RyanTheOptimist I figured out the cause of the timeout error. It's the combination of QUICHE's expiration time of QUIC Datagram in its Datagram queue, and Envoy using the `defer_send_in_response_to_packets_` of the QUIC connection config for CPU usage optimization.

I've made a temporary fix in this PR by setting a higher expiry time for QUIC Datagrams on the Envoy side, but the real fix requires the QUICHE library change, probably setting the default minimum expiry time higher. If you are fine with this temporary fix, I will send a subsequent PR for removing the temporary fix after this one gets merged, the QUICHE change is made, and rolled into Envoy. 

Looks like the CI is blocked due to the conflict in the `changelogs/current.yaml`. Could you resolve it? Thank you!

RyanTheOptimist commented:
> Thank you, @lizan.
> 
> @RyanTheOptimist I figured out the cause of the timeout error. It's the combination of QUICHE's expiration time of QUIC Datagram in its Datagram queue, and Envoy using the `defer_send_in_response_to_packets_` of the QUIC connection config for CPU usage optimization.
> 
> I've made a temporary fix in this PR by setting a higher expiry time for QUIC Datagrams on the Envoy side, but the real fix requires the QUICHE library change, probably setting the default minimum expiry time higher. If you are fine with this temporary fix, I will send a subsequent PR for removing the temporary fix after this one gets merged, the QUICHE change is made, and rolled into Envoy.
> 
> Looks like the CI is blocked due to the conflict in the `changelogs/current.yaml`. Could you resolve it? Thank you!

To fix the changelog you'll need to do a main merge.

jeongseokson commented:
/retest

RyanTheOptimist commented:
Looks like some tidy failures here:

https://dev.azure.com/cncf/envoy/_build/results?buildId=140504&view=logs&jobId=b7634614-24f3-5416-e791-4f3affaaed6c&j=b7634614-24f3-5416-e791-4f3affaaed6c&t=d565ee52-b888-5d80-7fea-dfaed104c390

Also coverage:

Code coverage for source/extensions/upstreams is lower than limit of 96.6 (93.8)
Code coverage for source/extensions/upstreams/http is lower than limit of 96.6 (93.1)
Code coverage for source/extensions/upstreams/http/generic is lower than limit of 96.6 (84.6)
Code coverage for source/extensions/upstreams/http/udp is lower than limit of 96.6 (78.0)

Detailed coverage results here:
https://storage.googleapis.com/envoy-pr/e9f2ada/coverage/index.html

RyanTheOptimist commented:
/wait
on CI

jeongseokson commented:
Hi @RyanTheOptimist 

I added more test cases to increase the test coverage.

Because of the coverage bug of excluding nested brackets of case statements, it's not possible to reach the coverage threshold in extensions/upstreams/http/generic/config.cc (https://storage.googleapis.com/envoy-pr/e9f2ada/coverage/source/extensions/upstreams/http/generic/config.cc.gcov.html). The maximum coverage I can achieve is 88.5.

I added it to the exception list to make it pass the coverage test. I set the threshold as 85.0 to give some headroom. Let me know if you have any comments or questions. Thank you!

jeongseokson commented:
/retest
